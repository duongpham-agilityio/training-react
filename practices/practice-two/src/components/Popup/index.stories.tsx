import { Meta, Story } from '@storybook/react';

// Component
import { Popup, PopupProps } from 'components';

export const Template: Story<PopupProps> = () => (
  <Popup
    title="Are you sure to deleted this book"
    description="This action can not undo, so please careful with this action"
  />
);

export default {
  title: 'Practice/components/Popup',
  component: Popup,
} as Meta;
