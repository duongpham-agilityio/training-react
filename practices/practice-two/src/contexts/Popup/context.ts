import { createContext } from 'react';

export type PopupState = {
  dispatch: (_fc: () => void) => void;
};

export const PopupContext = createContext<PopupState | null>(null);
