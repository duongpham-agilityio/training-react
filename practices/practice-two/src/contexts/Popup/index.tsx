import { ReactNode, MouseEvent, useState, useMemo, useCallback } from 'react';

// Components
import { Popup } from 'components';

// Contexts
import { PopupContext } from 'contexts/Popup/context';

// Constants
import { MESSAGES } from '@constants';

// const initialState = {
//   isShown: false,
//   accepter: () => {},
// };

const PopupProvider = ({ children }: { children: ReactNode }) => {
  const [state, setState] = useState<{
    isShown: boolean;
    accepter: (_event: MouseEvent) => void;
  } | null>(null);

  const props = useMemo(() => {
    return {
      dispatch: (callback: () => void) => {
        setState({
          isShown: true,
          accepter: callback,
        });
      },
    };
  }, []);

  const onAccept = useCallback(
    (event: MouseEvent) => {
      state?.accepter(event);

      setState(null);
    },
    [state]
  );

  return (
    <PopupContext.Provider value={props}>
      {children}

      {state?.isShown && (
        <Popup
          title={MESSAGES.DELETE_TITLE}
          description={MESSAGES.DELETE_DESCRIPTION}
          onCancel={() => setState(null)}
          onAccept={onAccept}
        />
      )}
    </PopupContext.Provider>
  );
};

export default PopupProvider;
