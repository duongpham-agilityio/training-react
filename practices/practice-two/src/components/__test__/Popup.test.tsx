import { render } from '@testing-library/react';

// Components
import Popup, { PopupProps } from 'components/Popup';

const setup = (props: PopupProps) => render(<Popup {...props} />);

describe('Popup component', () => {
  it('Match to snapshot', () => {
    const { container } = setup({});

    expect(container).toMatchSnapshot();
  });
});
