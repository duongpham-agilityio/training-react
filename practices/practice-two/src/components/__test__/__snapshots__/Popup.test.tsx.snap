// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup component Match to snapshot 1`] = `
<div>
  <section>
    <div>
      <h2
        class="undefined undefined "
      />
      <p />
      <div>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          Cancel
        </button>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          Ok
        </button>
      </div>
    </div>
  </section>
</div>
`;
